docs/dev/migrating-to-v1.rst

```diff
@@ -107,7 +105,85 @@ with their equivalent functions
 Model
 -----
 
-Making the model region it's own component
+Moving from an inheritance to composition structure for the Model class
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+**Rationale**
+
+Prior to v1, the `Model` class was the only real place where developers could
+modify the behavior of Core through either subclassing it, or using various
+`Mixin` classes. All parts of a model were implemented as class properties
+forcing every model to use the same terminology. While this was enough for
+some users, it was too restrictive for others. For example, the SFINCS
+plugin uses multiple grids for its computation, which was not possible in
+the setup pre-v1. There was also a lot of code duplication for the use of
+several parts of a model such as `maps`, `forcing` and `states`. To offer
+users more modularity and flexibility, as well as improve maintainability, we
+have decided to move the core to a component based architecture rather than
+an inheritance based one.
+
+**Changes required**
+
+Here we will describe the specific changes needed to use a `Model` object.
+The changes necessary to have core recognize your plugins are described below.
+Now a `Model` is made up of several `Component` classes to which it can delegate work.
+While it should still be responsible for workloads that span multiple components
+it should delegate work to components whenever possible. For specific changes needed
+for appropriate components see their entry in this migration guide, but general
+changes will be described here.
+
+Implementing Model Components
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Components are objects that the `Model` class can delegate work to. Typically, they are associated with one object such as a grid,
+forcing or tables. To be able to work within a `Model` class properly a component must implement the following methods:
+
+- `read`: reading the component and its data from disk.
+- `write`: write the component in its current state to disk in the provided root.
+
+Additionally, it is highly recommended to also provide the following methods:
+
+- `set`: provide the ability to override the current data in the component.
```

Comment:
We should recommend lazy initialization as you mention with this `_initialize` function, but it is up to the plugin developer to implement this correctly. We cannot enforce anything.

Just like the `set` function. From hydromt-core, we will never call `set`, because it is a function to override data on a component. It is up to the plugin developer to implement this in a proper way.